_posts/2023-09-06-load-balancer.md

```diff
@@ -0,0 +1,906 @@
+
+```scala
+import cats.effect.{IO, IOApp}
+import cats.implicits.{catsSyntaxTuple2Parallel, catsSyntaxTuple2Semigroupal}
+import com.comcast.ip4s.{Host, Port}
+import com.ghurtchu.loadbalancer.domain.{Config, Urls}
+import com.ghurtchu.loadbalancer.domain.Backends.{Backends, HealthChecks}
+import com.ghurtchu.loadbalancer.http.HttpServer
+import com.ghurtchu.loadbalancer.services.{ParseUri, RoundRobin, UpdateBackendsAndGet}
+import pureconfig._
+import pureconfig.generic.auto._
+
+object Main extends IOApp.Simple {
+
+  override def run: IO[Unit] =
+    for {
+      config                   <- IO(ConfigSource.default.loadOrThrow[Config])
+      (backends, healthChecks) <- refs(config.backends)
+      (host, port)             <- IO.fromEither {
+        hostAndPort(
+          config.hostOr(fallback = "0.0.0.0"),
+          config.portOr(fallback = 8080),
+        )
+      }
+      _                        <- IO.println(s"Starting server on URL: $host:$port")
+      _                        <- HttpServer.start(
+        backends,
+        healthChecks,
+        port,
+        host,
+        config.healthCheckInterval,
+        ParseUri.impl,
+        UpdateBackendsAndGet.impl,
+        RoundRobin.forBackends,
+        RoundRobin.forHealthChecks,
+      )
+    } yield ()
+
+  private def refs(urls: Urls): IO[(Backends, HealthChecks)] =
+    (
+      IO.ref(urls),
+      IO.ref(urls),
+    ).parMapN { case (backendsRef, healthChecksRef) =>
+      (Backends(backendsRef), HealthChecks(healthChecksRef))
+    }
+
+  private def hostAndPort(
+    host: String,
+    port: Int,
+  ): Either[Config.InvalidConfig, (Host, Port)] =
+    (
+      Host.fromString(host),
+      Port.fromInt(port),
+    ).tupled
+      .toRight(Config.InvalidConfig)
+}
+```
+
+Almost done!
+
+Before testing the load balancer manually we need to have some backends.
+
+For that I wrote really simple Python flask application which looks like this:
+
+```python
+from flask import Flask
+import sys
+
+app = Flask(__name__)
+host = "localhost"
+port = int(sys.argv[1])
+
+@app.route("/")
+def hello():
+    return f"hello from {url(host, port)}"
+
+@app.route("/health")
+def health():
+    return f"{url(host, port)} is alive and healthy ^_^"
+
+def url(host, port):
+    return f"http://{host}:{port}"
+
+if __name__ == '__main__':
+    app.run(host=host, port=port, debug=True)
+```
+
+Now we can create two shell scripts for running load balancer and backends easily.
+
+We can write a shell script - `be`, with which we can run the python flask app:
+```shell
+#!/bin/sh
+
+python3 echoer.py $1
+```
+
+and a separate shell script - `lb`, for running the load balancer:
+```shell
+#!/bin/sh
+
+scala lb.jar $1
+```
+
+In this video you can see the live testing of load balancer:
+{% include video id="SkQ6s_nwCgY" provider="youtube" %}
+
+If you want to see the whole project you can view _[Source code](https://github.com/Ghurtchu/lb)_
+
+Thank you for your time!
+
```

Comment:
Add a conclusion chapter to the article summing up what you've done. Include your GitHub repo here